### PR TITLE
Throttle total resources used for environment maps

### PR DESCRIPTION
**Pull request by ggetz**
# Description

[As demonstrated in this Sandcastle example](https://sandcastle.cesium.com/#c=vVXdc9MwDP9XdH3KWOYk++CjdDtgwMEdO8YKvFAevERtvDp2z3b2wa7/O7LTbEkoPLKHTJZ+kn6SZTXXyjq4FniDBo5B4Q2cohV1xb4HXTQb5eF8qpXjQqGZjWK4nykAoeb6jb4dw5xLi7FXWZSYO6HVR1WInDttetaSF/rGjsGZutXoWhavlai4wwf9euflTM1UQ4rZHBWyAi/rxbTUN+8Nr9Ceo5lirlVBnL1XcJjXKiSH3CDFO9MFyqg2MoYSxaJ0MfBK18qRT7bT1LDJgcoJJ9Ayg5W+xtdSRp4DwFwbiCQ6EOSUvqR/k00Qknd3d+7Bo/zfA/KqQV51kFc9ZA+9bNDLDnq5QXfxebimlbYiFHjcXtIpN44krg7Y3OjqLS4Moo26rgB72f4BS58dHj7NXsAu1fAEUpamaRb3cYeHLD1KD56lTwl19TdU00tCLAmRpT1r07Q+6RJ5IdTikfMZdyVz+oLUXNkoOzja5rYSLi9Dc/60GS3lX0zlajDGH5r05z7cBflFGz5xkyEOwbYR0EbQWPB+v78aokyXV1lWDgJ/qWnmjCJ41N5T7Ok0wzyMHibujgIPJ5AXRXTf77iiiR+DH+S+vk0zfpAGgE4J4+5hAKv8QxnDICtQRrEtLSQJVILebF2di1uUU/GL6GX7z4c4qPitR01zLgmxT8OUDjDr3nndv4dNa5zh+RKLd23HmtZ1kOtWDAJ91r7jm1tc+YItuf0g46ZEh7duDLPRNyWFA9phsAcZ7bXGSvCwxmh1tQsl2nnsTne5zEaMJVNerSS+5Y4noZM2oYgh9IPAFtLN/eLM6EnRt62zKT98vfAPekeQ15dY/EeSR1tI/rOB6X+nmKVbG/kzvLgpV0XOrZPo39RXreUlN2eo6mgzE953FI8m1t1JPGlpvRLVShvn5z4iUg6JFPG0yWVNU+hYbm2bdJJ0XSeFuAZRHG/5wYRccmvJMq9leC6z0ckkIfwfrlKHpfL5Go3kdx5WZiefGiVjbJLQcbunawocRP4N), it's possible for the environment map generation to puts a large load on browser resources when many models are added in the same frame, and that can cause the browser to hang and potentially crash.

This fix ensure that the environment map computations are distributed across several frames once they hit an upper limit, preventing the large simultaneous load on browser resources. 

## Issue number and link

N/A

## Testing plan

1. Try the above Sandcastle at each various option. Even at 10 x 10 grids, frame rate should still be interactive.
1. Make sure that otherwise, environment maps are being populated, such as in the "glTF PBR Extensions" example

# Author checklist

- [x] I have submitted a Contributor License Agreement
- [x] I have added my name to `CONTRIBUTORS.md`
- [x] I have updated `CHANGES.md` with a short summary of my change
- [ ] I have added or updated unit tests to ensure consistent code coverage
- [ ] ~I have updated the inline documentation, and included code examples where relevant~
- [x] I have performed a self-review of my code

### CODE REVIEW

**Comment by ggetz** [comment]
@jjhembd Are you available to review?

**Comment by javagl** [comment]
I tried this out, and for the "10 cubed" case, it _seems_ to work at the first glance. prints the usual
`CesiumWidget.js:50 [Violation] 'requestAnimationFrame' handler took 56ms`
which is not a big deal. But after printing that ~50 times, it says
```
...
CesiumWidget.js:50 [Violation] 'requestAnimationFrame' handler took 57ms
CesiumWidget.js:50 [Violation] 'requestAnimationFrame' handler took 51ms
CesiumWidget.js:50 [Violation] 'requestAnimationFrame' handler took 62ms
WebGL: CONTEXT_LOST_WEBGL: loseContext: context lost
An error occurred while rendering.  Rendering has stopped.
DeveloperError: Expected width to be greater than 0, actual value was 0
Error
    at new DeveloperError (http://localhost:8080/Build/CesiumUnminified/index.js:8786:11)
    at Check.typeOf.number.greaterThan (http://localhost:8080/Build/CesiumUnminified/index.js:8865:11)
    at Texture (http://localhost:8080/Build/CesiumUnminified/index.js:36496:31)
    at FramebufferManager.update (http://localhost:8080/Build/CesiumUnminified/index.js:41320:34)
    at GlobeDepth.update (http://localhost:8080/Build/CesiumUnminified/index.js:201636:29)
    at updateAndClearFramebuffers (http://localhost:8080/Build/CesiumUnminified/index.js:221608:21)
    at Scene4.updateAndExecuteCommands (http://localhost:8080/Build/CesiumUnminified/index.js:221242:3)
    at render (http://localhost:8080/Build/CesiumUnminified/index.js:221993:9)
    at tryAndCatchError (http://localhost:8080/Build/CesiumUnminified/index.js:222007:5)
    at Scene4.render (http://localhost:8080/Build/CesiumUnminified/index.js:222062:5)
```
Maybe someone can confirm or report something else.

**Comment by ggetz** [comment]
Thanks for checking @javagl! Perhaps we need a different way of determining the maximum number of commands per frame. If you have the time, you could try setting `DynamicEnvironmentMapManager._maximumComputeCommandCount` on [this line](https://github.com/CesiumGS/cesium/pull/12320/files#diff-c87b236b76b3329e8116710a7b9e3415805be3eb22c1993c0022eedd25d6c95fR821-R822) to a lower value.

**Comment by javagl** [comment]
I have to emphasize that I don't even have a vague idea what ~"distributing 'browser resources' over multiple frames" means here (which 'resources' actually _have_ to be 'distributed' in that manner?). But from the symptoms, it looks like this _could_ be a true memory leak that either 1. kicks in in a single frame or 2. (after this PR) kicks in after multiple frames. Specifically, I'll probably have a look at that in the context of https://community.cesium.com/t/procedural-ibl-gpu-memory-leak/36760 and try to see if anything is "leaked" here (regardless of the number of frames that it is distributed over). Given that I'm not familiar with that part of the code, it would only be a short debugstepping-pass, but I'll give it a short try.

**Comment by ggetz** [comment]
> it looks like this could be a true memory leak that either 1. kicks in in a single frame or 2. (after this PR) kicks in after multiple frames. Specifically, I'll probably have a look at that in the context of https://community.cesium.com/t/procedural-ibl-gpu-memory-leak/36760 and try to see if anything is "leaked" here (regardless of the number of frames that it is distributed over) 

I agree and I'm looking into that thread as well. I still think the change in this PR is good overall as it avoid running tons and tons of commands all in one frame.

**Comment by javagl** [comment]
I added some debug logs, and pragmatically plugged https://github.com/greggman/webgl-memory into the code, to get an idea where the "leak" might be. (That's a pretty useful library, by the way). I used a slightly adjusted sandcastle for these tests (see below), to select more meaningful numbers of objects for that test. 

I'll post condensed versions of the console logs here.
<sup>(Why is it impossible to copy-and-paste this log output? \*sigh\* - we're evolving, but backwards...)</sup>

The message that starts with `"Creating textures..."` is a log output that I inserted in [`updateSpecularMaps`](https://github.com/CesiumGS/cesium/blob/b17154b1b2388be98d3bd3bc00cb53c91b15a063/packages/engine/Source/Scene/DynamicEnvironmentMapManager.js#L577), after noticing that this is likely the culprit here (some details below)

#### One cube
```
Creating textures, mipmapLevels=10 for 6 faces, total: 60, base size 256x256
CubeMap.TOTAL_TEXTURES 2
info  
  memory: 
    buffer: 724
    drawingbuffer: 10469736
    renderbuffer: 13959648
    texture: 33784562
    total: 58214670
  resources: 
    buffer: 4
    framebuffer: 6
    program: 6
    query: 0
    renderbuffer: 2
    sampler: 0
    shader: 0
    sync: 0
    texture: 18
    transformFeedback: 0
    vertexArray: 2
```

#### 10 cubes
```
Creating textures, mipmapLevels=10 for 6 faces, total: 60, base size 256x256
(10 times...)
Creating textures, mipmapLevels=10 for 6 faces, total: 60, base size 256x256
CubeMap.TOTAL_TEXTURES 11
info  
  memory: 
    buffer: 2572
    drawingbuffer: 10469736
    renderbuffer: 13959648
    texture: 188975090
    total: 213407046
  resources: 
    buffer: 26
    framebuffer: 7
    program: 8
    query: 0
    renderbuffer: 2
    sampler: 0
    shader: 0
    sync: 0
    texture: 631
    transformFeedback: 0
    vertexArray: 22
```

#### 100 cubes
```
Creating textures, mipmapLevels=10 for 6 faces, total: 60, base size 256x256
(100 times)
Creating textures, mipmapLevels=10 for 6 faces, total: 60, base size 256x256
CubeMap.TOTAL_TEXTURES 101
info
  memory: 
    buffer: 19372
    drawingbuffer: 10469736
    renderbuffer: 13959648
    texture: 1698936050
    total: 1723384806
  resources: 
    buffer: 226
    framebuffer: 7
    program: 8
    query: 0
    renderbuffer: 2
    sampler: 0
    shader: 0
    sync: 0
    texture: 6211
    transformFeedback: 0
    vertexArray: 212

```

### Summary:

The number of 'texture' objects that are allocated is basically `numberOfObjects * 60` (!). This is due to the line linked above: There are 10 mipmap levels for each of the 6 sides of the cubemaps. Each of these maps has a top-level size of 256x256.

The total size of the allocated texture memory is
- For 1 object: 33784562 (33 MB)
- For 10 objects: 188975090 (188 MB)
- For 100 objects: 1698936050: (~1.7 Giga(!)bytes)

### Conclusion

It just needs that much memory 🤷‍♂️ 

But from what I've grasped from looking over the code quickly, it _looks_ like the `_specularMapTextures` are no longer required after that `ComputeCommand` is executed. I'm not 100% sure about that. But one related thought was that it _might_ be possible to get rid of some of the textures sooner than later. A quick attempt was to insert
```javascript
      // XXX
      console.log("Destroy specular textures");
      const length = manager._specularMapTextures.length;
      for (let i = 0; i < length; ++i) {
        manager._specularMapTextures[i] =
        manager._specularMapTextures[i] && manager._specularMapTextures[i].destroy();
      }
      // XXX
```
at https://github.com/CesiumGS/cesium/blob/b17154b1b2388be98d3bd3bc00cb53c91b15a063/packages/engine/Source/Scene/DynamicEnvironmentMapManager.js#L572 (because it looked like this could be a place after which these textures are actually no longer required). This did seem to have _some_ effect, but not solve it completely. Maybe someone can more easily (and confidently) identify (further) points where some textures could be detroyed eagerly, to free up memory.


---

The sandcastle

```javascript
const viewer = new Cesium.Viewer("cesiumContainer", {
  globe: false,
  infoBox: false,
  selectionIndicator: false,
  shouldAnimate: true,
  skyBox: false
});

viewer.scene.debugShowFramesPerSecond = true;

function createModel(url, height, amount = 1) {
  viewer.entities.removeAll();
  for (let i = 0; i < amount; i++){ 
      const position = Cesium.Cartesian3.fromDegrees(
        -123.0744619 + i * 0.0001,
        44.0503706,
        height,
      );
      const heading = Cesium.Math.toRadians(135);
      const pitch = 0;
      const roll = 0;
      const hpr = new Cesium.HeadingPitchRoll(heading, pitch, roll);
      const orientation = Cesium.Transforms.headingPitchRollQuaternion(position, hpr);

      const entity = viewer.entities.add({
        name: url,
        position: position,
        orientation: orientation,
        model: {
          uri: url,
         // minimumPixelSize: 128,
          maximumScale: 20000,
        },
      });
      viewer.trackedEntity = entity;
    }
}

const options = [
 {
    text: "Unlit Box - 1",
    onselect: function () {
      createModel("../../SampleData/models/BoxUnlit/BoxUnlit.gltf", 10.0, 1);
    },
  },
  {
    text: "Unlit Box - 10",
    onselect: function () {
      createModel("../../SampleData/models/BoxUnlit/BoxUnlit.gltf", 10.0, 10);
    },
  },
  {
    text: "Unlit Box - 100",
    onselect: function () {
      createModel("../../SampleData/models/BoxUnlit/BoxUnlit.gltf", 10.0, 100);
    },
  },
  {
    text: "Unlit Box - 1000",
    onselect: function () {
      createModel("../../SampleData/models/BoxUnlit/BoxUnlit.gltf", 10.0, 1000);
    },
  },
];

Sandcastle.addToolbarMenu(options);
```

**Review by jjhembd — APPROVED** [review]
Thanks @ggetz! The code change makes sense to me.

As for performance, I don't notice a framerate difference on my machine. However, in `main` the "10 cubed" example crashes my browser. In this branch "10 cubed" runs fine.

File: packages/engine/Source/Scene/DynamicEnvironmentMapManager.js [hunk 1]
```diff
@@ -295,6 +300,59 @@ Object.defineProperties(DynamicEnvironmentMapManager.prototype, {
...
+ * @private
+ * @param {FrameState} frameState The current frame state
+ */
+DynamicEnvironmentMapManager._updateCommandQueue = (frameState) => {
+  if (
+    DynamicEnvironmentMapManager._nextFrameCommandQueue.length > 0 &&
+    DynamicEnvironmentMapManager._activeComputeCommandCount <
+      DynamicEnvironmentMapManager._maximumComputeCommandCount
+  ) {
+    let command = DynamicEnvironmentMapManager._nextFrameCommandQueue.pop();
```
[the remark this answers is not in the document]
Comment (ggetz) [reply packages/engine/Source/Scene/DynamicEnvironmentMapManager.js]: You're totally right @jjhembd. I pushed up a fix.

**Comment by ggetz** [comment]
Thanks @jjhembd! This should be ready for another look.

**Comment by jjhembd** [comment]
Looks good, thanks @ggetz!

**Comment by javagl** [comment]
> In this branch "10 cubed" runs fine.

I'm a bit surprised. For me, it also did **seem** to run fine, but after ~50 frames (each printing `requestAnimationFrame handler took ~50ms`) it _did_ crash. I'll probably try again...

**Comment by ggetz** [comment]
@javagl I'm hoping the remaining issues you see here are addressed in https://github.com/CesiumGS/cesium/pull/12370.

**Comment by javagl** [comment]
Yes, I left a short comment a thttps://github.com/CesiumGS/cesium/pull/12370#issuecomment-2537302740

**Pull request merged**